Mathlib/Algebra/Exact.lean

```diff
@@ -35,30 +36,106 @@ variable (f : M → N) (g : N → P)
 
 variable {f : M →ₗ[R] N} {g : N →ₗ[R] P}
 
 lemma Exact.linearMap_ker_eq (hfg : Exact f g) : ker g = range f :=
   SetLike.ext hfg
 
-lemma LinearMap.exact_iff : Exact f g ↔ LinearMap.ker g = LinearMap.range f :=
+lemma _root_.LinearMap.exact_iff :
+    Exact f g ↔ LinearMap.ker g = LinearMap.range f :=
   Iff.symm <| SetLike.ext_iff
 
+lemma _root_.LinearMap.exact_of_comp_eq_zero_of_ker_le_range
+    (h1 : g ∘ₗ f = 0) (h2 : ker g ≤ range f) : Exact f g :=
+  Exact.of_comp_eq_zero_of_eq_zero_imp_mem_range (congrArg DFunLike.coe h1) h2
+
+lemma _root_.LinearMap.exact_of_comp_eq_zero_of_eq_zero_imp_mem_range
+    (h1 : g ∘ₗ f = 0) (h2 : ∀ x, g x = 0 → x ∈ range f) : Exact f g :=
+  exact_of_comp_eq_zero_of_ker_le_range h1 h2
+
 lemma Exact.linearMap_comp_eq_zero (h : Exact f g) : g.comp f = 0 :=
   DFunLike.coe_injective h.comp_eq_zero
 
+lemma Surjective.precomp_exact_iff_exact {p : M' →ₗ[R] M} (h : Surjective p) :
+    Exact (f ∘ₗ p) g ↔ Exact f g :=
+  iff_of_eq <| forall_congr fun x =>
+    congrArg (g x = 0 ↔ x ∈ ·) (h.range_comp f)
+
+lemma Injective.postcomp_exact_iff_exact {i : P →ₗ[R] P'} (h : Injective i) :
+    Exact f (i ∘ₗ g) ↔ Exact f g :=
+  forall_congr' fun _ => iff_congr (LinearMap.map_eq_zero_iff _ h) Iff.rfl
+
+lemma _root_.LinearEquiv.conj_exact_iff_exact (e : N ≃ₗ[R] N') :
+    Exact (e ∘ₗ f) (g ∘ₗ (e.symm : N' →ₗ[R] N)) ↔ Exact f g := by
+  simp_rw [exact_iff, ker_comp, ← e.map_eq_comap, range_comp]
+  exact (Submodule.map_injective_of_injective e.injective).eq_iff
+
+variable
+    {f₁₂ : M →ₗ[R] N} {f₂₃ : N →ₗ[R] P} {g₁₂ : M' →ₗ[R] N'}
+    {g₂₃ : N' →ₗ[R] P'} {e₁ : M ≃ₗ[R] M'} {e₂ : N ≃ₗ[R] N'} {e₃ : P ≃ₗ[R] P'}
+
+lemma Exact.of_ladder_linear_equiv_of_exact
+    (h₁₂ : g₁₂ ∘ₗ e₁ = e₂ ∘ₗ f₁₂) (h₂₃ : g₂₃ ∘ₗ e₂ = e₃ ∘ₗ f₂₃)
+    (H : Exact f₁₂ f₂₃) : Exact g₁₂ g₂₃ := by
+  rw [← LinearEquiv.eq_comp_toLinearMap_symm] at h₁₂ h₂₃
+  rwa [h₁₂, h₂₃, comp_assoc, LinearEquiv.conj_exact_iff_exact,
+    e₁.symm.surjective.precomp_exact_iff_exact,
+    e₃.injective.postcomp_exact_iff_exact]
+
+lemma Exact.iff_of_ladder_linear_equiv
+    (h₁₂ : g₁₂ ∘ₗ e₁ = e₂ ∘ₗ f₁₂) (h₂₃ : g₂₃ ∘ₗ e₂ = e₃ ∘ₗ f₂₃) :
+    Exact g₁₂ g₂₃ ↔ Exact f₁₂ f₂₃ where
+  mp := have h₂₁ := (e₂.eq_toLinearMap_symm_comp (f₁₂ ∘ₗ e₁.symm) g₁₂).mpr <|
+          Eq.trans (comp_assoc _ _ _).symm <|
+            (e₁.comp_toLinearMap_symm_eq g₁₂ (e₂ ∘ₗ f₁₂)).mpr h₁₂.symm
+        have h₃₂ := (e₃.eq_toLinearMap_symm_comp (f₂₃ ∘ₗ e₂.symm) g₂₃).mpr <|
+          Eq.trans (comp_assoc _ _ _).symm <|
+            (e₂.comp_toLinearMap_symm_eq g₂₃ (e₃ ∘ₗ f₂₃)).mpr h₂₃.symm
+        of_ladder_linear_equiv_of_exact h₂₁ h₃₂
+  mpr := of_ladder_linear_equiv_of_exact h₁₂ h₂₃
+
 end LinearMap
 
+section Ring
+
+open LinearMap Submodule
+
+variable [Ring R] [AddCommGroup M] [AddCommGroup N] [AddCommGroup P]
+    [Module R M] [Module R N] [Module R P]
+
+lemma Exact.descends_to_quotient_iff {f : M →ₗ[R] N} {g : N →ₗ[R] P}
```

Comment:
Maybe `exact_mapQ_iff`? And if the name is for searchability we may put it in the docstring instead.